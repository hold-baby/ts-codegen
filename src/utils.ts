import { camelCase, Dictionary, find, indexOf, map, takeRight, trimEnd } from "lodash";
import prettier from "prettier";
import { CustomSchema } from "src/__types__/types";
import { ERROR_MESSAGES } from "src/constants";

export const toCapitalCase = (str?: string): string => {
  if (!str) {
    return "";
  }
  const camelStr = camelCase(str);
  return `${camelStr.charAt(0).toUpperCase()}${camelStr.slice(1)}`;
};

export const isArray = (data: any) => Object.prototype.toString.call(data) === "[object Array]";

export const isNumberLike = (n: any) => {
  return !isNaN(parseFloat(n)) && isFinite(n);
};

export const prettifyCode = (code: string) =>
  prettier.format(code, {
    printWidth: 120,
    trailingComma: "all",
    arrowParens: "always",
    parser: "typescript",
  });

export const toTypes = (obj: Dictionary<any> | string) => {
  if (!obj) {
    return "";
  }
  const list = map<string, any>(obj, (v: any, k: string) => `${quoteKey(k)}: ${v};`);
  return (
    obj &&
    `{
        ${list.sort().join("\n")}
      }`
  );
};

export const quoteKey = (k: string) => {
  const isOptional = indexOf(k, "?") > -1;
  return `'${trimEnd(k, "?")}'${isOptional ? "?" : ""}`;
};

export function testJSON(
  str: unknown,
  errorMsg: string = ERROR_MESSAGES.INVALID_JSON_FILE_ERROR,
  output: (message: string) => void = console.error,
) {
  if (typeof str !== "string") {
    return;
  }

  try {
    return JSON.parse(str);
  } catch (e) {
    output(errorMsg);
    return;
  }
}

export const setDeprecated = (operationId: string = "") =>
  `
  /**
  * @deprecated ${operationId}
  */
  `;

export const getUseExtends = (schemas: CustomSchema) =>
  !!find(schemas, (schema) => schema.$ref) && !!find(schemas, (schema) => schema.type == "object");

export const getRefId = (str?: string): string => {
  if (!str) {
    return "";
  }
  const list = str.split("/");
  return list[list.length - 1];
};

export const withRequiredName = (name: string, required?: boolean) => `${name}${required ? "" : "?"}`;

export const getPathsFromRef = (str?: string): string[] => {
  if (!str) {
    return [];
  }

  const paths = str.replace(/^#\//, "").split("/");
  return takeRight(paths, 2);
};

export const getFilename = (basePath?: string) =>
  basePath ? `./${basePath.split("/").join(".").slice(1)}` : "./api.client";
