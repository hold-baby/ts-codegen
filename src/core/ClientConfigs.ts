import { camelCase, chain, compact, map, pick } from "lodash";
import { getRefId } from "src/core/utils";
import { Register } from "src/core/Register";
import { CustomOperation, CustomParameter, CustomParameters, CustomPath, CustomReference } from "src/core/types";

export const getOperations = (path: CustomPath) =>
  pick(path, ["get", "post", "put", "delete", "patch", "head"]) as { [method: string]: CustomOperation };

export const pickParams = (params?: CustomParameters) => (type: "path" | "query" | "body" | "formData") => {
  const list = map(params, (param: CustomParameter | CustomReference) => {
    let data = param;

    if ((param as CustomReference).$ref) {
      const name = getRefId((param as CustomReference).$ref);
      data = Register.parameters[name];
    }

    if ((data as CustomParameter).in === type) {
      return data;
    }
  });

  return compact(list);
};

export const getRequestURL = (pathName: string, basePath?: string) => {
  const isPathParam = (str: string) => str.startsWith("{");
  const resolvedPathName = chain(pathName)
    .split("/")
    .map((p) => (isPathParam(p) ? `$${p}` : p))
    .join("/")
    .value();

  return `${basePath}${resolvedPathName === "/" && !!basePath ? "" : resolvedPathName}`;
};

export const getOperationId = (operationId?: string) => camelCase(operationId);
