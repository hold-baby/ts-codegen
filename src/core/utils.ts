import { camelCase, chain, Dictionary, find, forEach, indexOf, map, replace, some, trimEnd } from "lodash";
import prettier from "prettier";
import { ERROR_MESSAGES } from "src/core/constants";
import { CustomSchema } from "src/core/Type";

export const toCapitalCase = (str?: string): string => {
  if (!str) {
    return "";
  }
  const camelStr = camelCase(str);
  return `${camelStr.charAt(0).toUpperCase()}${camelStr.slice(1)}`;
};

export const addPrefix = (prefix: string) => (str: string = "") => `${prefix}${str}`;
export const addSuffix = (suffix: string) => (str: string = "") => `${str}${suffix}`;

export const addPrefixForInterface = addPrefix("I");
export const addPrefixForType = addPrefix("T");

export const arrayToObject = (arr: any[] = []) => {
  let obj: any = {};
  forEach(arr, (item) => {
    obj[item] = item;
  });
  return obj;
};

export const isArray = (data: any) => Object.prototype.toString.call(data) === "[object Array]";
export const isObject = (data: any) => Object.prototype.toString.call(data) === "[object Object]";
export const isNumber = (n: any) => {
  return !isNaN(parseFloat(n)) && isFinite(n);
};

export const prettifyCode = (code: string) =>
  prettier.format(code, {
    printWidth: 120,
    trailingComma: "all",
    arrowParens: "always",
    parser: "typescript",
  });

const ENUM_SUFFIX = `#EnumSuffix`;

export const toTypes = (obj: Dictionary<any> | string) => {
  if (!obj) {
    return "";
  }
  const list = map<string, any>(obj, (v: any, k: string) => `${quoteKey(k)}: ${replace(v, ENUM_SUFFIX, "")};`);
  return (
    obj &&
    `{
        ${list.sort().join("\n")}
      }`
  );
};

export const quoteKey = (k: string) => {
  const isOptional = indexOf(k, "?") > -1;
  return `'${trimEnd(k, "?")}'${isOptional ? "?" : ""}`;
};

export function testJSON(
  str: unknown,
  errorMsg: string = ERROR_MESSAGES.INVALID_JSON_FILE_ERROR,
  output: (message: string) => void = console.error,
) {
  if (typeof str !== "string") {
    return;
  }

  try {
    return JSON.parse(str);
  } catch (e) {
    output(errorMsg);
    return;
  }
}

export const toArrayType = (customType?: string) => (customType ? `${customType}[]` : "Array<any>");

export const toType = (builtInType: string = ""): string => {
  if (builtInType === "integer") {
    return "number";
  }

  return builtInType;
};

export const generateEnumType = (p = "", k = "") => `${toCapitalCase(p)}${toCapitalCase(k)}${ENUM_SUFFIX}`;

export const handleEnums = (enums: string[], enumName: string) => {
  const hasNumber = some(enums, (v) => isNumber(v));
  return hasNumber
    ? `export type ${enumName} = ${enums.map((item: string | number) => JSON.stringify(item)).join("|")}`
    : `export enum ${enumName} ${JSON.stringify(arrayToObject(enums)).replace(/:/gi, "=")}`;
};

export function generateEnums(data: Dictionary<any>, key: string) {
  if (!data) {
    return "";
  }

  const enums = data[key];
  const enumName = replace(key, ENUM_SUFFIX, "");
  return handleEnums(enums, enumName);
}

export const setDeprecated = (operationId: string = "") =>
  `
  /**
  * @deprecated ${operationId}
  */
  `;

export const getRequestURL = (pathName: string, basePath?: string) => {
  const isPathParam = (str: string) => str.startsWith("{");
  const resolvedPathName = chain(pathName)
    .split("/")
    .map((p) => (isPathParam(p) ? `$${p}` : p))
    .join("/")
    .value();

  return `${basePath}${resolvedPathName === "/" && !!basePath ? "" : resolvedPathName}`;
};

export const getUseExtends = (schemas: CustomSchema) =>
  !!find(schemas, (schema) => schema.$ref) && !!find(schemas, (schema) => schema.type == "object");
