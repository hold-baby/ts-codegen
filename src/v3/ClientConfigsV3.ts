import { Schema } from "src/core/Schema";
import { chain, filter, get, isEmpty, keys, map, pick, reduce } from "lodash";
import { IClientConfigs } from "src/core/types";
import { IOperation, IPathItem, IPaths, IReference, IRequestBody, IResponse, TParameter } from "src/v3/OpenAPI";
import { CustomSchema, CustomType } from "src/core/Type";
import { toCapitalCase } from "src";

export class ClientConfigsV3 {
  clientConfigs: IClientConfigs[] = [];
  schemaHandler: Schema;

  static of(paths: IPaths, basePath: string = "") {
    return new ClientConfigsV3(paths, basePath);
  }

  constructor(private paths: IPaths, private basePath: string) {
    this.schemaHandler = new Schema();
    this.clientConfigs = reduce(
      this.paths,
      (configs: IClientConfigs[], path: IPathItem, pathName: string) => [
        ...configs,
        ...this.buildConfig(path, pathName),
      ],
      [],
    );
  }

  buildConfig(path: IPathItem, pathName: string) {
    const operations = pick(path, ["get", "post", "put", "delete", "patch", "head"]) as { [k: string]: IOperation };

    return keys(operations).map((method) => {
      const operation = operations[method];
      const pickParamsByType = this.pickParams(operation.parameters);
      const pathParams = pickParamsByType("path") as Array<TParameter | IReference>;
      const queryParams = pickParamsByType("query") as Array<TParameter | IReference>;
      const getParamTypes = this.getParamTypes(operation.operationId);
      const getNames = (list: any[]) => (isEmpty(list) ? [] : map(list, (item) => item.name));

      return {
        url: this.getUrl(this.basePath, pathName),
        method,
        operationId: operation.operationId, //camelCase(operation.operationId) TODO: add tests later, 向后兼容？
        TResp: this.getResponseType(operation.responses),
        TReq: {
          ...getParamTypes(pathParams),
          ...getParamTypes(queryParams),
          ...this.getBodyParamsTypes(operation.requestBody),
        },
        pathParams: getNames(pathParams),
        queryParams: getNames(queryParams),
        contentType: this.getContentType(operation.requestBody),
        deprecated: operation.deprecated,
      };
    });
  }

  getBodyParamsTypes(requestBody?: IReference | IRequestBody) {
    // TODO: will handle another content type later
    if (!requestBody) {
      return;
    }
    const schema =
      (requestBody as IRequestBody)?.content["application/json"]?.schema ||
      (requestBody as IRequestBody)?.content["application/x-www-form-urlencoded"]?.schema ||
      (requestBody as IRequestBody)?.content["multipart/form-data"]?.schema;

    return {
      requestBody: this.schemaHandler.convert(schema as CustomSchema),
    };
  }

  private getParamTypes = (_name?: string) => {
    return (params?: Array<TParameter | IReference>): { [key: string]: CustomType } | undefined => {
      if (!params) {
        return;
      }
      return params.reduce((results, param) => {
        //TODO: will handle reference here

        return {
          ...results,
          [`${(param as TParameter).name}${(param as TParameter).required ? "" : "?"}`]: this.schemaHandler.convert(
            get(param, "schema", param),
            `${toCapitalCase(_name)}${toCapitalCase((param as TParameter).name)}`,
          ),
        };
      }, {});
    };
  };

  getUrl(basePath: string, pathName: string) {
    const path = chain(pathName)
      .split("/")
      .map((p) => (this.isPathParam(p) ? `$${p}` : p))
      .join("/")
      .value();

    return `${basePath}${path === "/" && !!basePath ? "" : path}`;
  }

  isPathParam = (str: string) => str.startsWith("{");

  private getResponseType = (responses: IOperation["responses"]) => {
    const response200 = get(responses, "200");
    const response201 = get(responses, "201");

    if ((response200 as IReference)?.$ref || (response201 as IReference)?.$ref) {
      return this.schemaHandler.convert(response200 || response201);
    }

    const v = (response200 as IResponse)?.schema || (response201 as IResponse)?.schema;
    return v ? this.schemaHandler.convert(v) : undefined;
  };

  getContentType = (requestBody?: IRequestBody | IReference) => {
    if (!requestBody) {
      return "";
    }
    // TODO: handle reference later

    // TODO: handle other content type later
    return keys((requestBody as IRequestBody).content)[0];
  };

  pickParams = (params?: Array<TParameter | IReference>) => (type: "path" | "query" | "body" | "formData") => {
    if (!params) {
      return;
    }
    return filter(params, (param) => {
      // TODO: will handle $ref later
      if (param.$ref) {
        return;
      }
      return (param as TParameter).in === type;
    });
  };
}
