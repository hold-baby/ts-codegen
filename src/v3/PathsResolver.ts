import { chain, compact, Dictionary, filter, get, isEmpty, map, pick, reduce, sortBy } from "lodash";
import { IOperation, IPathItem, IPaths, IReference, IRequestBody, IResponse, TParameter } from "src/v3/OpenAPI";
import { SchemaResolver } from "src/core/SchemaResolver";
import {generateEnums, toTypes} from "src/core/utils";

interface IParams {
  pathParams: Array<TParameter | IReference>;
  queryParams: Array<TParameter | IReference>;
  cookieParams: Array<TParameter | IReference>;
}

interface IResolvedPath {
  url: string;
  method: string;
  TResp: any;
  TReq: any;
  operationId?: string;
  pathParams: string[];
  queryParams: string[];
  cookieParams: string[];
  requestBody: { required: boolean | undefined } | null;
}

// TODO: 1. 将 inline 的 requestParams 和 requestBody 抽成单独的 interface，方便外面使用
// TODO: 2. query 不要全部 ...，而是以具体的 {[key]: value} 形式，避免外部应用一些不需要的 query

export class PathsResolver {
  resolver: SchemaResolver;
  resolvedPaths: IResolvedPath[] = [];
  extraDefinitions: Dictionary<any> = {};

  static of(paths: IPaths, basePath: string = "") {
    return new PathsResolver(paths, basePath);
  }

  constructor(private paths: IPaths, private basePath: string) {
    this.resolver = SchemaResolver.of((k, v) => {
      if (k) {
        this.extraDefinitions[k] = v;
      }
    });
  }

  scan = () => {
    this.resolvedPaths = reduce(
      this.paths,
      (results: IResolvedPath[], pathItem: IPathItem, pathName: string) => [
        ...results,
        ...this.resolve(pathItem, pathName),
      ],
      [],
    );
    return this;
  };

  toRequest = (): string[] => {
    const data = sortBy(this.resolvedPaths, (o) => o.operationId);
    const requests = data.map((v: IResolvedPath) => {
      const TReq = !isEmpty(v.TReq) ? toTypes(v.TReq) : undefined;
      const requestParamList = [
        ...v.pathParams,
        ...v.queryParams,
        ...v.cookieParams,
        v.requestBody ? "requestBody" : undefined,
      ];
      const params = this.toRequestParams(get(v, "queryParams"));

      return `export const ${v.operationId} = createRequestAction<${TReq}, ${v.TResp}>('${v.operationId}', (${
        !isEmpty(requestParamList) ? `${this.toRequestParams(requestParamList)}` : ""
      }) => ({url: \`${v.url}\`, method: "${v.method}", ${v.requestBody ? `data: requestBody,` : ""}${
        params ? `params: ${params},` : ""
      }}));`;
    });

    const enums = Object.keys(this.extraDefinitions).map((k) => generateEnums(this.extraDefinitions, k));
    return [...requests, ...enums];
  };

  resolve(pathItem: IPathItem, pathName: string) {
    const operations = pick(pathItem, ["get", "post", "put", "delete", "patch", "head"]);

    return Object.keys(operations).map((method) => {
      const operation = (operations as Dictionary<IOperation>)[method];
      const params = this.getAllParams(operation);

      return {
        url: this.getUrl(this.basePath, pathName),
        method,
        operationId: operation.operationId,
        TReq: this.getRequestTypes(params, operation.requestBody),
        TResp: this.getResponseTypes(operation.responses),
        ...this.getParamsNames(params),
        requestBody: this.getRequestBody(operation.requestBody),
      };
    });
  }

  getUrl = (basePath: string, pathName: string) => {
    const path = chain(pathName)
      .split("/")
      .map((p) => (this.isPathParam(p) ? `$${p}` : p))
      .join("/")
      .value();

    return `${basePath}${path === "/" && !!basePath ? "" : path}`;
  };

  toRequestParams = (data: any[] = []) =>
    !isEmpty(data)
      ? `{
    ${compact(data).join(",\n")}
    }`
      : undefined;

  isPathParam = (str: string) => str.startsWith("{");

  // TODO: handle the case when v.parameters = Reference
  getAllParams = (o: IOperation) => {
    // TODO: when parameters has enum
    const getParams = (type: "path" | "query" | "cookie") => {
      return filter(o.parameters, (param) => {
        // TODO: handle $ref here
        if (param.$ref) {
          return false;
        }
        return (param as TParameter).in === type;
      });
    };

    return {
      pathParams: getParams("path"),
      queryParams: getParams("query"),
      cookieParams: getParams("cookie"),
    };
  };

  getParamsNames = (params: IParams) => {
    const getNames = (list: any[]) => (isEmpty(list) ? [] : map(list, (item) => item.name));
    return {
      pathParams: getNames(params.pathParams),
      queryParams: getNames(params.queryParams),
      cookieParams: [], // TODO: will handle cookieParams later
    };
  };

  getRequestTypes = (params: IParams, requestBody?: IReference | IRequestBody) => ({
    ...this.getPathParamsTypes(params.pathParams),
    ...this.getQueryParamsTypes(params.queryParams),
    ...this.getRequestBodyTypes(requestBody),
  });

  getPathParamsTypes = (pathParams: Array<TParameter | IReference>) => {
    return pathParams.reduce(
      (results, param) => ({
        ...results,
        [`${(param as TParameter).name}${(param as TParameter).required ? "" : "?"}`]: this.resolver.toType({
          ...(param as TParameter).schema,
          _name: (param as TParameter).name,
          _propKey: (param as TParameter).name,
        }),
      }),
      {},
    );
  };

  getQueryParamsTypes = (queryParams: Array<TParameter | IReference>) =>
    queryParams.reduce(
      (o, v) => ({
        ...o,
        [`${(v as TParameter).name}${(v as TParameter).required ? "" : "?"}`]: this.resolver.toType({
          ...(v as TParameter).schema,
          _name: (v as TParameter).name,
          _propKey: (v as TParameter).name,
        }),
      }),
      {},
    );

  getCookieParamsTypes = (_: Array<TParameter | IReference>) => {};

  getRequestBodyTypes = (requestBody?: IReference | IRequestBody) => {
    if (!requestBody) {
      return "";
    }

    return {
      requestBody: this.resolver.toType((requestBody as IRequestBody).content["application/json"].schema),
    };
  };

  // TODO: handle Response or Reference
  // TODO: responses.200.schema.type ==="array"
  // TODO: response.200.schema.type ==="object" (additionalProperties)
  // TODO: response.200.schema.type==="string" | "number" | "integer" | "boolean"
  // TODO: responses.200.headers 存在时
  // TODO: responses.201 同上

  getResponseTypes = (responses: { [responseName: string]: IResponse | IReference }) => {
    return this.resolver.toType(
      get(responses, [200, "content", "application/json", "schema"]) ||
        get(responses, [201, "content", "application/json", "schema"]),
    );
  };

  getRequestBody = (requestBody?: IReference | IRequestBody) => {
    if (!requestBody) {
      return null;
    }
    // TODO: handle IReference
    if (requestBody.$ref) {
      return null;
    }
    return {
      required: (requestBody as IRequestBody).required,
    };
  };
}
