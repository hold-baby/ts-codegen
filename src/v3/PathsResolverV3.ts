import { chain, compact, Dictionary, filter, get, isEmpty, map, pick, reduce, sortBy, isString } from "lodash";
import { IOperation, IPathItem, IPaths, IReference, IRequestBody, IResponse, TParameter } from "src/v3/OpenAPI";
import { SchemaHandler } from "src/core/SchemaHandler";
import { generateEnums, toTypes } from "src/core/utils";
import { resolve } from "src/core/ReusableTypes";

interface IParams {
  pathParams: Array<TParameter | IReference>;
  queryParams: Array<TParameter | IReference>;
  cookieParams: Array<TParameter | IReference>;
}

interface IResolvedPath {
  url: string;
  method: string;
  TResp: any;
  TReq: any;
  operationId?: string;
  pathParams: string[];
  queryParams: string[];
  cookieParams: string[];
  requestBody: { required: boolean | undefined } | null;
}

// TODO: 1. 将 inline 的 requestParams 和 requestBody 抽成单独的 interface，方便外面使用
// TODO: 2. query 不要全部 ...，而是以具体的 {[key]: value} 形式，避免外部应用一些不需要的 query

export class PathsResolverV3 {
  schemaHandler: SchemaHandler;
  resolvedPaths: IResolvedPath[] = [];
  extraDefinitions: Dictionary<any> = {};

  static of(paths: IPaths, basePath: string = "", reusableSchemas: Dictionary<any>) {
    return new PathsResolverV3(paths, basePath, reusableSchemas);
  }

  constructor(private paths: IPaths, private basePath: string, private reusableSchemas: Dictionary<any>) {
    this.schemaHandler = SchemaHandler.of((k, v) => {
      if (k) {
        this.extraDefinitions[k] = v;
      }
    });
  }

  scan = () => {
    this.resolvedPaths = reduce(
      this.paths,
      (results: IResolvedPath[], pathItem: IPathItem, pathName: string) => [
        ...results,
        ...this.resolve(pathItem, pathName),
      ],
      [],
    );
    return this;
  };

  toRequest = (): string[] => {
    const data = sortBy(this.resolvedPaths, (o) => o.operationId);
    const requests = data.map((v: IResolvedPath) => {
      const TReq = !isEmpty(v.TReq) ? toTypes(v.TReq) : undefined;
      const requestParamList = [
        ...v.pathParams,
        ...v.queryParams,
        ...v.cookieParams,
        v.requestBody ? "requestBody" : undefined,
      ];
      const params = this.toRequestParams(get(v, "queryParams"));
      const operationId = v.operationId?.replace(/\s/gi, "");
      return `export const ${operationId} = createRequestAction<${TReq}, ${v.TResp}>('${operationId}', (${
        !isEmpty(requestParamList) ? `${this.toRequestParams(requestParamList)}` : ""
      }) => ({url: \`${v.url}\`, method: "${v.method}", ${v.requestBody ? `data: requestBody,` : ""}${
        params ? `params: ${params},` : ""
      }}));`;
    });

    const enums = Object.keys(this.extraDefinitions).map((k) => generateEnums(this.extraDefinitions, k));
    return [...requests, ...enums];
  };

  resolve(pathItem: IPathItem, pathName: string) {
    const operations = pick(pathItem, ["get", "post", "put", "delete", "patch", "head"]);

    return Object.keys(operations).map((method) => {
      const operation = (operations as Dictionary<IOperation>)[method];
      const params = this.getAllParams(operation);

      return {
        url: this.getUrl(this.basePath, pathName),
        method,
        operationId: operation.operationId,
        TReq: this.getRequestTypes(params, operation.requestBody, operation.operationId),
        TResp: this.getResponseTypes(operation.responses),
        ...this.getParamsNames(params),
        requestBody: this.getRequestBody(operation.requestBody),
      };
    });
  }

  getUrl = (basePath: string, pathName: string) => {
    const path = chain(pathName)
      .split("/")
      .map((p) => (this.isPathParam(p) ? `$${p}` : p))
      .join("/")
      .value();

    return `${basePath}${path === "/" && !!basePath ? "" : path}`;
  };

  toRequestParams = (data: any[] = []) => {
    const inputs = compact(data);
    return !isEmpty(inputs)
      ? `{
    ${inputs.join(",\n")}
    }`
      : "";
  };

  isPathParam = (str: string) => str.startsWith("{");

  // TODO: handle the case when v.parameters = Reference
  getAllParams = (o: IOperation) => {
    // TODO: when parameters has enum
    const getParams = (type: "path" | "query" | "cookie") => {
      return filter(o.parameters, (param) => {
        // TODO: handle $ref here
        if (param.$ref) {
          return false;
        }
        return (param as TParameter).in === type;
      });
    };

    return {
      pathParams: getParams("path"),
      queryParams: getParams("query"),
      cookieParams: getParams("cookie"),
    };
  };

  getParamsNames = (params: IParams) => {
    const getNames = (list: any[]) => (isEmpty(list) ? [] : map(list, (item) => item.name));
    return {
      pathParams: getNames(params.pathParams),
      queryParams: getNames(params.queryParams),
      cookieParams: [], // TODO: will handle cookieParams later
    };
  };

  getRequestTypes = (params: IParams, requestBody?: IReference | IRequestBody, _name?: string) => ({
    ...this.getPathParamsTypes(params.pathParams, _name),
    ...this.getQueryParamsTypes(params.queryParams, _name),
    ...this.getRequestBodyTypes(requestBody, _name),
  });

  getPathParamsTypes = (pathParams: Array<TParameter | IReference>, _name?: string) => {
    return pathParams.reduce(
      (results, param) => ({
        ...results,
        [`${(param as TParameter).name}${(param as TParameter).required ? "" : "?"}`]: resolve(
          this.schemaHandler.toType({
            ...(param as TParameter).schema,
            _name,
            _propKey: (param as TParameter).name,
          }),
          this.reusableSchemas,
        ),
      }),
      {},
    );
  };

  getQueryParamsTypes = (queryParams: Array<TParameter | IReference>, _name?: string) =>
    queryParams.reduce(
      (o, v) => ({
        ...o,
        [`${(v as TParameter).name}${(v as TParameter).required ? "" : "?"}`]: resolve(
          this.schemaHandler.toType({
            ...(v as TParameter).schema,
            _name,
            _propKey: (v as TParameter).name,
          }),
          this.reusableSchemas,
        ),
      }),
      {},
    );

  getCookieParamsTypes = (_: Array<TParameter | IReference>) => {};

  getRequestBodyTypes = (requestBody?: IReference | IRequestBody, _name?: string) => {
    if (!requestBody) {
      return "";
    }
    const rb = resolve(
      this.schemaHandler.toType({
        ...((requestBody as IRequestBody).content["application/json"]?.schema ||
          (requestBody as IRequestBody).content["application/x-www-form-urlencoded"]?.schema ||
          (requestBody as IRequestBody).content["multipart/form-data"]?.schema),
        _name,
      }),
      this.reusableSchemas,
    );

    return {
      requestBody: isString(rb) ? rb : toTypes(rb),
    };
  };

  // TODO: handle Response or Reference
  // TODO: responses.200.schema.type ==="array"
  // TODO: response.200.schema.type ==="object" (additionalProperties)
  // TODO: response.200.schema.type==="string" | "number" | "integer" | "boolean"
  // TODO: responses.200.headers 存在时
  // TODO: responses.201 同上

  getResponseTypes = (responses: { [responseName: string]: IResponse | IReference }) => {
    return resolve(
      this.schemaHandler.toType(
        get(responses, [200, "content", "application/json", "schema"]) ||
          get(responses, [201, "content", "application/json", "schema"]),
      ),
      this.reusableSchemas,
    );
  };

  getRequestBody = (requestBody?: IReference | IRequestBody) => {
    if (!requestBody) {
      return null;
    }
    // TODO: handle IReference
    if (requestBody.$ref) {
      return null;
    }
    return {
      required: (requestBody as IRequestBody).required,
    };
  };
}
