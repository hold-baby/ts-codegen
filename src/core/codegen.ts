import { PathsResolverV3 } from "src/v3/PathsResolverV3";
import { ReusableTypes } from "src/core/ReusableTypes";
import { prettifyCode, testJSON } from "src/core/utils";
import axios from "axios";
import { map } from "lodash";
import { ERROR_MESSAGES } from "src/core/constants";
import * as fs from "fs";
import * as path from "path";
import { IOpenAPI } from "src/v3/OpenAPI";
import { Spec } from "swagger-schema-official";
import { PathsResolverV2 } from "src";

export const codegen = (spec: IOpenAPI | Spec): string => {
  if (!spec) {
    return "";
  }

  if ((spec as IOpenAPI).openapi) {
    return [
      ...PathsResolverV3.of(spec.paths, spec.basePath)
        .scan()
        .toRequest(),
      ...ReusableTypes.of(spec).gen(),
    ].join("\n\n");
  }

  return [
    ...PathsResolverV2.of(spec.paths, spec.basePath)
      .scan()
      .toRequest(),
    ...ReusableTypes.of(spec).gen(),
  ].join("\n\n");
};

const getFilename = (basePath?: string) =>
  basePath
    ? `./${basePath
        .split("/")
        .join(".")
        .slice(1)}`
    : "./api.client";

const write = (output: string, filename: string, str: string) => {
  if (!fs.existsSync(output)) {
    fs.mkdirSync(output);
  }

  fs.writeFileSync(path.resolve(output, `./${filename}.ts`), prettifyCode(str), "utf-8");
};

// const codegenFromRemote = () => {};

export const fetchSwaggerJSON = (clients: string[] = [], timeout: number = 10 * 1000) => {
  const instance = axios.create({
    timeout,
  });

  return Promise.all(
    map(clients, (client) => {
      instance
        .get(client)
        .then((response) => response.data)
        .catch((error) => {
          console.error(`${error.code}: ${ERROR_MESSAGES.FETCH_CLIENT_FAILED_ERROR}`);
        });
    }),
  );
};

export const codegenFromConfig = () => {
  const codegenConfigPath = path.resolve("ts-codegen.config.json");

  const getCodegenConfig = () =>
    fs.existsSync(codegenConfigPath)
      ? require(codegenConfigPath)
      : {
          output: ".output",
          actionCreatorImport: "",
          clients: [],
        };
  const { output, actionCreatorImport, timeout, data, clients } = getCodegenConfig();

  const writeSpecToFile = (spec: IOpenAPI | Spec) => {
    const fileStr = `${actionCreatorImport} ${codegen(spec)}`;
    write(output, getFilename(spec.basePath), fileStr);
  };

  (data || []).map((file: string) => {
    const specStr = fs.readFileSync(file, "utf8");
    const spec = testJSON(specStr);

    writeSpecToFile(spec);
  });

  if (clients) {
    fetchSwaggerJSON(clients, timeout).then((results: any[]) => {
      results.forEach((spec: IOpenAPI | Spec) => {
        writeSpecToFile(spec);
      });
    });
  }
};

codegenFromConfig();
