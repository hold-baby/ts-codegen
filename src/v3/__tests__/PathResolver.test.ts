import { PathsResolverV3 } from "src/v3/PathsResolverV3";
import swagger from "examples/swagger.v3.petstore.expanded.json";
import { prettifyCode } from "src/core/utils";

describe("PathsResolver", () => {
  it("should get resolved paths by swagger schema", () => {
    expect(PathsResolverV3.of((swagger as any).paths, "/api/test").scan().resolvedPaths).toEqual(
      expectedPathResolvedData,
    );
  });

  it("should generate correct request from swagger data", () => {
    const fileStr = PathsResolverV3.of((swagger as any).paths, "/api/test")
      .scan()
      .toRequest();
    expect(prettifyCode(fileStr.join("\n\n"))).toMatchSnapshot();
  });
});

const expectedPathResolvedData = [
  {
    TReq: {
      "limit?": "number",
      "tags?": "string[]",
    },
    TResp: "Pet[]",
    method: "get",
    operationId: "findPets",
    cookieParams: [],
    pathParams: [],
    queryParams: ["tags", "limit"],
    url: "/api/test/pets",
    requestBody: null,
  },
  {
    TReq: {
      requestBody: "NewPet",
    },
    TResp: "Pet",
    method: "post",
    operationId: "addPet",
    cookieParams: [],
    pathParams: [],
    queryParams: [],
    url: "/api/test/pets",
    requestBody: {
      required: true,
    },
  },
  {
    TReq: {
      id: "number",
    },
    TResp: "Pet",
    method: "get",
    operationId: "findPetById",
    cookieParams: [],
    pathParams: ["id"],
    queryParams: [],
    url: "/api/test/pets/${id}",
    requestBody: null,
  },
  {
    TReq: {
      id: "number",
    },
    TResp: "",
    method: "delete",
    operationId: "deletePet",
    cookieParams: [],
    pathParams: ["id"],
    queryParams: [],
    url: "/api/test/pets/${id}",
    requestBody: null,
  },
];
