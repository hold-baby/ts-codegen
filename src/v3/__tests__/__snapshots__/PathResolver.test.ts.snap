// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathsResolver should generate correct request from swagger data 1`] = `
"export const addPet = createRequestAction<
  {
    requestBody: NewPet;
  },
  Pet
>(\\"addPet\\", ({ requestBody }) => ({ url: \`/api/test/pets\`, method: \\"post\\", data: requestBody }));

export const deletePet = createRequestAction<{
  id: number;
}>(\\"deletePet\\", ({ id }) => ({ url: \`/api/test/pets/\${id}\`, method: \\"delete\\" }));

export const findPetById = createRequestAction<
  {
    id: number;
  },
  Pet
>(\\"findPetById\\", ({ id }) => ({ url: \`/api/test/pets/\${id}\`, method: \\"get\\" }));

export const findPets = createRequestAction<
  {
    limit?: number;
    tags?: string[];
  },
  Pet[]
>(\\"findPets\\", ({ tags, limit }) => ({
  url: \`/api/test/pets\`,
  method: \\"get\\",
  params: {
    tags,
    limit,
  },
}));
"
`;
